### ISSUE TITLE
Unresponsive Gmail and other JS interfaces

### ISSUE THREAD

**Issue by dmik** [bitwiseworks/mozilla-os2#248]
FF 45 exposes performance degradation in some usage scenarios. One of them is gmail.com UI. If you log in, then start composing a new message and then press the Tab key to traverse through input fields, you will get a system freeze when the focus traversal is about to loopback (move from the last input field to the first one). The freeze is quite hard — it gives a 100% CPU load and also entirely blocks the PM message queue making the whole system unresponsive. Usually unfreezes in about 30-60 seconds.

There are reports of other JS UIs behaving the same way.

Note that this is new in FF 45 (so looks like a regression of JS changes in there). There is #179 but it is related to FF 38 and describes a different issue (which I guess has gone now).

**Comment by dmik** [comment]
Note https://github.com/bitwiseworks/mozilla-os2/issues/19#issuecomment-361917012. Now I think that the problem here is not JS itself (especially after having run all JS tests within #247 and finding no test that would reveal any such problem), but rather PM that somehow gets flooded with too many messages or such. What gives it this stress is still a question.

**Comment by dmik** [comment]
Got a thought after #258 that we might get this strange behavior because it is expected upstream that OMTC is enabled on all supported platforms so non-OMTC builds are not even properly tested any more. And some parts of Firefox might closely rely on OMTC and not function well otherwise.

**Comment by dmik** [comment]
Also, absence of e10s on OS/2 (#228) might also somehow cause this.

**Comment by dmik** [comment]
Investigating event loops within #266 and #265 reveals that when Gmail becomes unresponsive, any message loop spinning stops. So this unresponsiveness is *not* caused by a massive message flood as I once guessed (see https://github.com/bitwiseworks/mozilla-os2/issues/266#issuecomment-388361800). Still it looks like some JS processing takes extremely long on the UI thread and it can't be interrupted to process UI messages despite all the means JS has for that. Either our port misses something which I can't find or those means I ported (i.e. interrupting by injecting an invalid opcode to cause an exception) does not work as expected by the JS developers (there are some limitations, see #253 for details).

However, I did find some change between esr38 and esr45 that affects this problem in some way. It's not directly related to JS, it's more about how UI events are handled in FF (the design is terrible to my taste, overly complex and too many layers of abstraction). This is the changeset https://hg.mozilla.org/integration/mozilla-inbound/rev/727955f60cd89654b981ad1da8783e9b9f02b9f4 where they changed the maximum non-native event starvation interval from 20ms to 10ms (i.e. the maximum time during which native UI events are handled before the event loop switches back to PL ("portable") events). Somehow reverting it back to 20ms gives a better balance between native and PL events and Gmail becomes less unresponsive in my tests.

Please test. For this you need http://rpm.netlabs.org/test/ff45_9_0_t5.7z and put a newer XUL from http://rpm.netlabs.org/test/x_t5_1.7z on top of it.

**Comment by dmik** [comment]
The fun thing is that the change of 20 ms to 10 ms in `nsBaseApp.cpp` from mentioned Mozilla changeset (which is still on the tip of the trunk BTW) looks like a parallel experiment of the developer committed by an accident. As the rest of the changeset doesn't seem to be related to it at all. Very funny considering the effect it had on our port.

**Comment by dmik** [comment]
Forgot to mention that @ydario and Elbert Pol report that the latest XUL.DLL with this starvation delay increased back to 20 ms makes Firefox behave more like esr38 again, i.e. it's still slower than on e.g. Windows, but much better than before. So I encourage everyone to test. This may be it.

**Comment by dryeo** [comment]
Didn't notice a huge improvement with Firefox, perhaps due to using a 
fairly new profile.
Building SeaMonkey and Thunderbird with the patched nsBaseAppShell.cpp 
does seem to help quite a bit, especially Thunderbird where getting 
focus can be slow with no CPU or disk activity, though I do notice that 
typing this seems slower.
Probably worth committing though I might try experimenting with 30 ms
Dave

**Comment by dmik** [comment]
@dryeo Well I played around with this value. Higher ones as well as lower ones make the situation worse. This should not have such an impact and this is why I say that the whole thingy of several layers of event handling (native events, PL events, Chromium events, all intermixed) is a terribly bad idea. This impact is timing-based and is perhaps less visible on other platforms due to their increased robustness (modern drivers with GPU acceleration, modern memory management, modern task scheduler etc etc).

I feel like I commit 20 ms and mark this as resolved.

**Comment by dryeo** [comment]
I tested Thunderbird with 30 and yes, performance dropped so 20 seems 
like the sweet spot.

**Comment by dspiatkowski** [comment]
@dmik and @dryeo My results here produce a system that, while more responsive and better behaved, is quite honestly nowhere near the behaviour of FF38. It is better, no doubt, but strangely enough, what Dave posted in another ticket, namely the use of a different THEME is actually of much greater effect. For me, the 'Past Modern revisited, reloaded' is what appears to work.

Given that I had literally re-built my profile in preparation for the FF45 stuff I do not think the profile itself is a problem, however, I will create a TEST profile and will try that out to see.

To summarize, the latest changes do make things better, but it's a tiny difference here.

**Comment by dmik** [comment]
@dspiatkowski While I believe a different theme might change the behavior when minimized, I don't really think that it has something to do with Gmail unresponsiveness.

**Comment by dryeo** [comment]
I was seeing the same slowdowns in SeaMonkey, unlike the heavy CPU when 
minimized. Not sure if it is the latest build or the new Gmail interface 
but things are better now.
Note I don't usually use the web mail interface on Gmail so didn't test 
the new interface lately until yesterday.

**Comment by guzzi-g5** [comment]
Tested with gmail, youtube. Seems to work ok, much more stable, although some sound stuttering after playing about an hour of clips. After closing all memory is released, that's an improvement. Scrolling a long way down on facebook still causes unresponsiveness here but much less and faster recovery than the previous version.

**Comment by dmik** [comment]
Okay, thanks everybody for testing. I consider the current result as acceptable. I think that this issue should be closed as we can't do more for it at the moment.
